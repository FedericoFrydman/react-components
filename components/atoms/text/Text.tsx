import { forwardRef, ReactElement, Ref } from 'react';
import cn from 'classnames';

import './text.scss';
import '../../assets/styles/utilities.scss';

export type TextAs = 'p' | 'span';

export type TextSize = 'body' | 'description' | 'tiny' | 'small' | 'medium' | 'large' | 'extra-large';

export type TextWeight = 'normal' | 'semibold' | 'bold';

export type TextColor = 'N1000' | 'N800' | 'N700' | 'N500' | 'N300' | 'N100' | 'N00';

export interface TextProps extends Omit<React.HTMLProps<HTMLParagraphElement>, 'css'> {
  /**
   * Render a span or a paragraph. Default 'span'.
   */
  as?: TextAs;
  /**
   * Text value to be rendered
   */
  text?: string;
  /**
   * Defines the text height
   */
  textHeight?: string;
  /**
   * Size of the text. Defaults to `medium`.
   */
  textSize?: TextSize;
  /**
   * Weight of the text
   */
  textWeight?: TextWeight;
  /**
   * Sets the `textWeight` to bold
   * @deprecated Use `fontWeight='bold'` instead. Will be removed in future verison.
   */
  bold?: boolean;
  /**
   * Sets the `textWeight` to semibold
   * @deprecated Use `fontWeight='semibold'` instead. Will be removed in future verison.
   */
  semiBold?: boolean;
  /**
   * Sets the text style to italic.
   */
  italic?: boolean;
  /**
   * Add a strikethrough to the text.
   */
  strike?: boolean;
  /**
   * Text value to be rendered
   */
  children?: string | number | ReactElement;
  /**
   * Custom classes for the component.
   */
  classNames?: cn.Argument;
  /**
   * Color of the text
   * @deprecated Text defaults to N800 color. This prop will be removed in a future verison.
   */
  color?: TextColor;
  /**
   * Modifies the text as separator
   */
  separator?: boolean;
  /**
   * The ref to the HTML DOM element.
   */
  ref?: Ref<HTMLParagraphElement>;
}

const TextComponent = (
  {
    as,
    bold,
    children,
    color,
    italic,
    semiBold,
    separator,
    strike,
    text,
    textHeight,
    textSize = 'medium',
    textWeight = 'normal',
    classNames = [],
    ...props
  }: TextProps,
  ref: TextProps['ref'],
) => {
  // Warn if bold or semibold props are used
  if (bold || semiBold) {
    console.warn('bold & semibold are deprecated. Please use `textWeight` instead.');
  }

  // Warn if color prop is used
  if (color) {
    console.warn('color is deprecated. Text is defaulted to $N800');
  }

  // Warn if body or description textSize values are used
  if (textSize === 'body' || textSize === 'description') {
    console.warn(
      'textSizes `body` and `description` have been deprecated. Please use `tiny`, `small`, `medium`, `large`, or `extra-large` instead.',
    );
  }

  // Set textWeight
  if (semiBold) {
    textWeight = 'semibold';
  }
  if (bold) {
    textWeight = 'bold';
  }

  const classText = 'lsux-text';
  const setHeight = { height: textHeight, overflow: 'auto' };
  const useHeight = textHeight ? setHeight : {};

  const cnText = cn(
    {
      [`${classText}`]: true,
      [`${classText}--separator`]: separator,
      [`${classText}--italic`]: italic,
      [`${classText}--strike`]: strike,
      [`${classText}--${textSize}`]: textSize,
      [`${classText}--${textWeight}`]: textWeight,
    },
    classNames,
  );

  if (as === 'span') {
    return (
      <span {...props} ref={ref} style={{ ...useHeight, ...props.style }} className={cnText}>
        {text ? text : children}
      </span>
    );
  } else {
    return (
      <p {...props} ref={ref} style={{ ...useHeight, ...props.style }} className={cnText}>
        {text ? text : children}
      </p>
    );
  }
};

export const Text = forwardRef<HTMLParagraphElement, TextProps>(TextComponent);
