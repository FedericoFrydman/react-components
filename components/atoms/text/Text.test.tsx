import { createRef } from 'react';
import { render, screen } from '@testing-library/react';
import { shallow, ShallowWrapper } from 'enzyme';
import cn from 'classnames';

import { Text, TextProps, TextSize, TextWeight } from './Text';

describe('Text Component', () => {
  const baseClass = `lsux-text`;
  let testId: string;
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    wrapper = shallow(<Text text="some text" classNames={['custom--class']} />);
    testId = 'test-text';
  });

  it('should render the component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('props', () => {
    describe('classNames', () => {
      describe('when provided classNames', () => {
        const classNames = ['my-class-1', 'my-class-2'];

        it('gives the text the provided classNames', () => {
          const text = renderText({ classNames });

          expect(text).toHaveClass(cn(classNames), { exact: false });
        });
      });
    });

    describe('text', () => {
      const text = 'testing text';

      it('is displayed when present', () => {
        renderText({ text });

        expect(screen.getByText(text)).toBeVisible();
      });
    });

    describe('children', () => {
      const childrenText = 'testing children';
      const children = <span>{childrenText}</span>;
      const childrenHTML = shallow(children).html();

      it('is renders the children when passed', () => {
        const wrapper = renderText({ children });
        expect(wrapper).toContainHTML(childrenHTML);
      });
    });

    describe('textHeight', () => {
      const text = 'testing text';
      describe('when it is defined', () => {
        it('is present on the text element', () => {
          renderText({ text, textHeight: '10px' });
          expect(screen.getByText(text)).toContainHTML('height: 10px;');
        });
      });
    });

    describe('textSize', () => {
      const textSizes: TextSize[] = ['body', 'description', 'tiny', 'small', 'medium', 'large', 'extra-large'];
      const text = 'Test my text size';
      let textSizeClass: string;

      describe('when "textSize" is not defined', () => {
        it('defaults "textSize" to the "body"', () => {
          renderText({ text });

          const labelEl = screen.getByText(text);
          expect(labelEl).toHaveClass(`${baseClass}--medium`);
        });

        function itAssignsTheRightClassForTextSize(textSize: TextSize) {
          describe(`when "textSize" is "${textSize}"`, () => {
            beforeEach(() => {
              textSizeClass = `${baseClass}--${textSize}`;
            });

            it(`assigns the element the ${textSizeClass} class`, () => {
              renderText({ text, textSize });

              const textEl = screen.getByText(text);
              expect(textEl).toHaveClass(textSizeClass);
            });
          });
        }

        textSizes.forEach((textSize) => itAssignsTheRightClassForTextSize(textSize));
      });
    });

    describe('textWeight', () => {
      const textWeights: TextWeight[] = ['normal', 'semibold', 'bold'];
      const text = 'Test my text weight';
      let textWeightClass: string;

      describe('when "textWeight" is not defined', () => {
        it('defaults "textWeight" to the "normal"', () => {
          renderText({ text });

          const labelEl = screen.getByText(text);
          expect(labelEl).toHaveClass(`${baseClass}--normal`);
        });

        function itAssignsTheRightClassForTextWeight(textWeight: TextWeight) {
          describe(`when "textWeight" is "${textWeight}"`, () => {
            beforeEach(() => {
              textWeightClass = `${baseClass}--${textWeight}`;
            });

            it(`assigns the element the ${textWeightClass} class`, () => {
              renderText({ text, textWeight });

              const textEl = screen.getByText(text);
              expect(textEl).toHaveClass(textWeightClass);
            });
          });
        }

        textWeights.forEach((textWeight) => itAssignsTheRightClassForTextWeight(textWeight));
      });
    });

    describe('bold', () => {
      const boldClass = `${baseClass}--bold`;
      let bold: boolean;

      describe('when bold is true', () => {
        beforeEach(() => (bold = true));

        it(`has the ${boldClass} class`, () => {
          expect(renderText({ bold })).toHaveClass(boldClass, { exact: false });
        });
      });

      describe('when bold is false', () => {
        beforeEach(() => (bold = false));

        it(`does not have the ${boldClass} class`, () => {
          expect(renderText({ bold })).not.toHaveClass(boldClass, { exact: false });
        });
      });
    });

    describe('semiBold', () => {
      const semiBoldClass = `${baseClass}--semibold`;
      let semiBold: boolean;

      describe('when semiBold is true', () => {
        beforeEach(() => (semiBold = true));

        it(`has the ${semiBold} class`, () => {
          expect(renderText({ semiBold })).toHaveClass(semiBoldClass, { exact: false });
        });
      });

      describe('when semiBold is false', () => {
        beforeEach(() => (semiBold = false));

        it(`does not have the ${semiBold} class`, () => {
          expect(renderText({ semiBold })).not.toHaveClass(semiBoldClass, { exact: false });
        });
      });
    });

    describe('italic', () => {
      const italicClass = `${baseClass}--italic`;
      let italic: boolean;

      describe('when italic is true', () => {
        beforeEach(() => (italic = true));

        it(`has the ${italicClass} class`, () => {
          expect(renderText({ italic })).toHaveClass(italicClass, { exact: false });
        });
      });

      describe('when italic is false', () => {
        beforeEach(() => (italic = false));

        it(`does not have the ${italicClass} class`, () => {
          expect(renderText({ italic })).not.toHaveClass(italicClass, { exact: false });
        });
      });
    });

    describe('strike', () => {
      const strikeClass = `${baseClass}--strike`;
      let strike: boolean;

      describe('when strike is true', () => {
        beforeEach(() => (strike = true));

        it(`has the ${strikeClass} class`, () => {
          expect(renderText({ strike })).toHaveClass(strikeClass, { exact: false });
        });
      });

      describe('when strike is false', () => {
        beforeEach(() => (strike = false));

        it(`does not have the ${strikeClass} class`, () => {
          expect(renderText({ strike })).not.toHaveClass(strikeClass, { exact: false });
        });
      });
    });
  });

  describe('ref', () => {
    const ref = createRef<HTMLParagraphElement>();

    it('provides a ref to the text', () => {
      expect(renderText({ ref })).toEqual(ref.current);
    });
  });

  const renderText = (props: TextProps = {}): HTMLElement => {
    render(<Text {...props} data-testid={testId} />);

    return screen.getByTestId(testId);
  };
});
