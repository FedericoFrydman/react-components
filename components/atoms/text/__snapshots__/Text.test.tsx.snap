// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text Component should render the component 1`] = `
<p
  className="lsux-text lsux-text--medium lsux-text--normal custom--class"
  style={Object {}}
>
  some text
</p>
`;
